ipa-core/src/ff/curve_points.rs

```diff
@@ -196,7 +196,55 @@ impl rand::distributions::Distribution<RP25519> for rand::distributions::Standar
     fn sample<R: crate::rand::Rng + ?Sized>(&self, rng: &mut R) -> RP25519 {
         let mut scalar_bytes = [0u8; 64];
         rng.fill_bytes(&mut scalar_bytes);
-        RP25519(RistrettoPoint::from_uniform_bytes(&scalar_bytes).compress())
+        RP25519(RistrettoPoint::from_uniform_bytes(&scalar_bytes).into())
+    }
+}
+
+/// Internal representation of Ristretto point, suitable
+/// for our needs.
+/// Due to constraints imposed by dalek crate,
+/// we can't construct a zero uncompressed Ristretto point
+/// at compile time. It is only possible to construct
+/// a compressed ristretto from 0 byte array in const context.
+/// We work around that limitation by adding an enum value
+/// that represents a zero value.
+#[derive(Clone, Copy, Eq, Debug)]
+enum RistrettoRepr {
+    /// In PRF code this path is never used as
+    /// we always construct Ristretto points from scalars
+    Zero,
+    Point(RistrettoPoint),
+}
+
+impl PartialEq for RistrettoRepr {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_point().eq(other.as_point())
+    }
+}
+
+impl RistrettoRepr {
+    #[inline]
+    fn zero() -> &'static RistrettoPoint {
+        static INSTANCE: OnceLock<RistrettoPoint> = OnceLock::new();
+        INSTANCE.get_or_init(|| {
+            let zero = CompressedRistretto([0_u8; 32]);
+            // we could also cache the compressed Ristretto, if we end up
+            // sending a lot of zeroes
+            zero.decompress().unwrap()
+        })
+    }
+
+    pub fn as_point(&self) -> &RistrettoPoint {
+        match self {
+            Self::Zero => Self::zero(),
```

Comment:
I'm inclined to panic here. The current implementation is not going to have constant-time operations when the repr is `Zero`. As you note, we don't create zero points now, so we are fine to use this implementation, but it would be safer to prevent someone trying to use it in a different context where it wouldn't be okay.

Reply:
Panicking seems like a safer option for sure. Your point about constant time operation is valid, but I think it also applies to our previous implementation. the cost of `Rp25519::ZERO + Rp25519::from(Scalar)` is not the same as `Rp25519::from(Scalar) + Rp25519::from(Scalar)`.